### PR TITLE
Custom messages

### PR DESCRIPTION
**Pull request by vosscodes**
Merging fuzzmz's PR #91 adding the option to include custom messages in freestyle projects with support for build parameters, environment variables, etc.

Added a listener/log method to the CardBuilder and a mock listener to the tests.

### CODE REVIEW

**Comment by vosscodes** [comment]
I'm not a java developer by any means, so please let me know if anything needs adjusting

**Comment by vosscodes** [comment]
@damianszczepanik I merged the latest master, do you need anything else from me to push this through?

File: src/main/java/jenkins/plugins/office365connector/Office365ConnectorWebhookNotifier.java [hunk 1]
```diff
@@ -80,10 +86,16 @@ public void sendBuildCompletedNotification() {
             return;
         }
 
-        Card card = cardBuilder.createCompletedCard();
+        Card card;
 
         for (Webhook webhook : webhooks) {
             if (decisionMaker.isStatusMatched(webhook) && decisionMaker.isAtLeastOneRuleMatched(webhook)) {
+                if (webhook.isIncludeCustomMessage()) {
```
Comment (damianszczepanik) [inline src/main/java/jenkins/plugins/office365connector/Office365ConnectorWebhookNotifier.java]: repeated code

File: src/main/java/jenkins/plugins/office365connector/Webhook.java [hunk 2]
```diff
@@ -158,6 +169,20 @@ public void setMacros(List<Macro> macros) {
         this.macros = Util.fixNull(macros);
     }
 
+    @DataBoundSetter
+    public void setIncludeCustomMessage(boolean includeCustomMessage) {
```
Comment (damianszczepanik) [inline src/main/java/jenkins/plugins/office365connector/Webhook.java]: why do we need to have extra boolean flag?

File: src/main/java/jenkins/plugins/office365connector/FactsBuilder.java [hunk 3]
```diff
@@ -118,6 +126,12 @@ private Collection sortUsers(Set<User> authors) {
                 .collect(Collectors.toList());
     }
 
+    public void addCustom(String message, Run run) throws IOException, InterruptedException {
+        EnvVars envVars;
+        envVars = run.getEnvironment(new LogTaskListener(LOGGER, Level.INFO));
+        addFact(ADDITIONAL_INFO, envVars.expand(message));
```
Comment (damianszczepanik) [inline src/main/java/jenkins/plugins/office365connector/FactsBuilder.java]: I think you can reuse method that processes macro which I believe produces the same results

File: src/main/java/jenkins/plugins/office365connector/CardBuilder.java [hunk 4]
```diff
@@ -55,7 +71,7 @@ public Card createStartedCard() {
         return card;
     }
 
-    public Card createCompletedCard() {
+    public Card createCompletedCard(String... messages) {
```
Comment (damianszczepanik) [inline src/main/java/jenkins/plugins/office365connector/CardBuilder.java]: do we have to pass messages to each method separately?

**Comment by damianszczepanik** [comment]
A few:
- test coverage should not be decreased
- there should be one integration test for custom message as new request https://github.com/jenkinsci/office-365-connector-plugin/tree/master/src/test/resources/requests to be tested
- I'd like to have possibility to define more than one custom message

**Comment by damianszczepanik** [comment]
Feature merged

**Comment by vosscodes** [comment]
thanks so much for taking care of this! I'm caught up with day job and a contract right now unfortunately

**Pull request closed without merging**